merino-web/src/extractors.rs

```diff
@@ -86,85 +84,29 @@ struct SuggestQuery {
 struct SupportedLanguagesWrapper(SupportedLanguages);
 
 impl FromRequest for SupportedLanguagesWrapper {
-    type Config = ();
     type Error = ActixError;
     type Future = Ready<Result<Self, Self::Error>>;
 
     fn from_request(req: &HttpRequest, _payload: &mut Payload) -> Self::Future {
-        /// Parse the quality value from a string of the form q=`<quality value>`.
-        fn parse_quality_value(quality_value: &str) -> Result<f64, HandlerError> {
-            let (_, weight_as_string) = quality_value
-                .split_once('=')
-                .ok_or(HandlerErrorKind::MalformedHeader("Accept-Language"))?;
-
-            let weight = weight_as_string
-                .parse::<f64>()
-                .map_err(|_| HandlerErrorKind::MalformedHeader("Accept-Language"))?;
-
-            if (0.0..=1.0).contains(&weight) {
-                Ok(weight)
-            } else {
-                Err(HandlerErrorKind::MalformedHeader("Accept-Language").into())
-            }
-        }
-
-        /// Parse the Accept-Language HTTP header.
-        fn parse_language(raw_language: &str) -> Result<Language, ActixError> {
-            let (locale_or_wildcard, quality_value) =
-                if let Some((language, quality_value)) = raw_language.split_once(';') {
-                    let quality_value = Some(parse_quality_value(quality_value)?);
-
-                    (language, quality_value)
-                } else {
-                    (raw_language, None)
-                };
-
-            let language = if locale_or_wildcard == "*" {
-                Language {
-                    language_identifier: LanguageIdentifier::Wildcard,
-                    quality_value,
-                }
-            } else if let Some((language, region)) = locale_or_wildcard.split_once("-") {
-                Language {
-                    language_identifier: LanguageIdentifier::Locale {
-                        language: language.to_lowercase(),
-                        region: Some(region.to_lowercase()),
-                    },
-                    quality_value,
+        future::ready({
+            if req.headers().contains_key("Accept-Language") {
+                match AcceptLanguage::parse(req) {
+                    Ok(languages) if languages.is_empty() => {
+                        // AcceptLanguage::parse() returns an empty Vec for certain types of
+                        // errors in the header. In these cases, we want to return an error.
+                        Err(HandlerErrorKind::MalformedHeader("Accept-Language").into())
+                    }
+                    Err(_) => Err(HandlerErrorKind::MalformedHeader("Accept-Language").into()),
```

Comment:
Returning an error from this function will abort the whole request with a 400 status code. Is that what we want here? I think it may be better to act as if the header wasn't there, and return `QualityItem::max(Preference::Any)`.

I don't think the [RFC for this header](https://httpwg.org/specs/rfc7231.html#rfc.section.5.3.5) specifies a behavior if the header is invalid, but it does include the following:

> If the header field is present in a request and none of the available representations for the response have a matching language tag, the origin server can either disregard the header field by treating the response as if it is not subject to content negotiation or honor the header field by sending a 406 (Not Acceptable) response. However, the latter is not encouraged, as doing so can prevent users from accessing content that they might be able to use (with translation software, for example).

that implies to me that we should err on the side of returning content instead of returning errors.

Reply:
Ah yeah, I agree. I'll make the change 👍🏻